### ISSUE TITLE
Problem migrating from DBN to lasagne NeuralNet: NaN for each epoch

### ISSUE THREAD

**Issue by bmilde** [dnouri/nolearn#35]
Lasagne looks fantastic, thanks for integrating it into nolearn! However, I have trouble transitioning from nolearn's DBN to the new lasagne NeuralNet.

Here is what happens:

Done loading and transforming data, traindata size:  83.5334777832 MB
Distribution of classes in train data:
[[  0.00000000e+00   5.82160000e+04]
 [  1.00000000e+00   5.12730000e+04]] 2
conf: momentum: 0.01 self.learn_rates: 0.01
fitting classifier... nolearn
  InputLayer            (None, 200)             produces     200 outputs
  DenseLayer            (None, 50)              produces      50 outputs
  DenseLayer            (None, 2)               produces       2 outputs

 Epoch  |  Train loss  |  Valid loss  |  Train / Val  |  Valid acc  |  Dur
--------|--------------|--------------|---------------|-------------|-------
     1  |         nan  |         nan  |          nan  |     45.05%  |  0.7s
     2  |         nan  |         nan  |          nan  |     46.47%  |  0.6s
     3  |         nan  |         nan  |          nan  |     45.77%  |  0.6s
     4  |         nan  |         nan  |          nan  |     47.06%  |  0.6s
     5  |         nan  |         nan  |          nan  |     47.07%  |  0.7s
     6  |         nan  |         nan  |          nan  |     47.06%  |  0.7s
     7  |         nan  |         nan  |          nan  |     47.08%  |  0.7s
     8  |         nan  |         nan  |          nan  |     53.71%  |  0.7s
     9  |         nan  |         nan  |          nan  |     47.05%  |  0.6s
    10  |         nan  |         nan  |          nan  |     47.05%  |  0.6s
    11  |         nan  |         nan  |          nan  |     47.05%  |  0.6s
    12  |         nan  |         nan  |          nan  |     47.05%  |  0.7s
    13  |         nan  |         nan  |          nan  |     47.05%  |  0.6s
    14  |         nan  |         nan  |          nan  |     47.05%  |  0.6s

I tried fiddling with different learning rates (1,0.1,0.01,... 0.0000001 even 0.0), momentum rates, different optimisers (sgd,nestrov, rmsprop ...every method that lasagne offers), input sizes, no. of hidden units, two and one hidden layer, all to no avail.

The mnist example from lasagne runs fine though.

Here is my DBN code, which also runs fine and produces models with >0.90% accuracy (on an audio gender detection task), on the same data:

                clf = DBN([X_train.shape[1], self.hid_layer_units, self.hid_layer_units, self._no_classes],
                        dropouts=self.dropouts,
                        learn_rates=self.learn_rates,
                        learn_rates_pretrain=self.learn_rates_pretrain,
                        minibatch_size=self.minibatch_size,
                        learn_rate_decays=self.learn_rate_decays,
                        learn_rate_minimums=self.learn_rate_minimums,
                        epochs_pretrain=self.pretrainepochs,
                        epochs=self.epochs,
                        momentum= self.momentum,
                        real_valued_vis=True,
                        use_re_lu=True,
                        verbose=1)

I've translated that into:

                clf = NeuralNet(
                        layers=[  # three layers: one hidden layer
                                ('input', layers.InputLayer),
                                ('hidden', layers.DenseLayer),
                                #('hidden', layers.DenseLayer),
                                ('output', layers.DenseLayer),
                                ],
                                # layer parameters:
                                input_shape=(None, X_train.shape[1]),  
                                hidden_num_units=self.hid_layer_units,  
                                output_num_units=self._no_classes,
                                output_nonlinearity=None,

                                eval_size=0.1,

                                # optimization method:
                                update=sgd,
                                update_learning_rate=self.learn_rates,
                                #update_momentum=momentum,

                                regression=False, 
                                max_epochs=self.epochs,  
                                verbose=1,
                                )

Is there anything obvious that I've missed here? How can I debug this?

**Comment by bmilde** [comment]
Apparently my y_train vector was uint8 instead of int32. Casting that array into int32 made SGD work, however I've haven't had luck with the more advanced optimisers yet.

**Comment by bmilde** [comment]
So apparently I get random behaviour with the same parameters (maybe different random validation splits?). Sometimes I get output similar to this, while on other runs it stays at nan:

 Epoch  |  Train loss  |  Valid loss  |  Train / Val  |  Valid acc  |  Dur
--------|--------------|--------------|---------------|-------------|-------
     1  |         nan  |         nan  |          nan  |     47.83%  |  0.6s
     2  |         nan  |   -4.046479  |          nan  |     52.17%  |  0.6s
     3  |   -4.317877  |   -4.494806  |     0.960637  |     52.17%  |  0.6s
     4  |   -4.617792  |   -4.706948  |     0.981059  |     52.17%  |  0.6s
     5  |   -4.800839  |   -4.868985  |     0.986004  |     52.17%  |  0.6s
     6  |   -4.950880  |   -5.009015  |     0.988394  |     52.17%  |  0.6s
     7  |   -5.083247  |   -5.134719  |     0.989976  |     52.17%  |  0.6s
     8  |   -5.203283  |   -5.249783  |     0.991143  |     52.17%  |  0.6s
     9  |   -5.313925  |   -5.356560  |     0.992041  |     52.17%  |  0.6s
    10  |   -5.417135  |   -5.456659  |     0.992757  |     52.17%  |  0.6s
    11  |   -5.514205  |   -5.551058  |     0.993361  |     52.17%  |  0.6s
    12  |   -5.605903  |   -5.640381  |     0.993887  |     52.17%  |  0.6s
    13  |   -5.692804  |   -5.725166  |     0.994347  |     52.17%  |  0.6s
    14  |   -5.775415  |   -5.805890  |     0.994751  |     52.17%  |  0.6s
    15  |   -5.854187  |   -5.882978  |     0.995106  |     52.17%  |  0.6s
    16  |   -5.929519  |   -5.956805  |     0.995419  |     52.17%  |  0.6s
    17  |   -6.001755  |   -6.027685  |     0.995698  |     52.17%  |  0.6s
    18  |   -6.071187  |   -6.095888  |     0.995948  |     52.17%  |  0.6s
    19  |   -6.138061  |   -6.161638  |     0.996174  |     52.17%  |  0.6s
    20  |   -6.202585  |   -6.225127  |     0.996379  |     52.17%  |  0.6s
    21  |   -6.264939  |   -6.286526  |     0.996566  |     52.17%  |  0.6s

Is there any restriction on the input that I've missed? Non-negative values or something similar? Should my training input be a multiple of batch_size?

**Comment by dnouri** [comment]
I think what you're missing is `output_nonlinearity=lasagne.nonlinearities.softmax`.  Sorry for a lack of proper documentation for this.  But there's an MNIST example included in the tests if you want to have a look: https://github.com/dnouri/nolearn/blob/master/nolearn/tests/test_lasagne.py#L41-L91

**Comment by bmilde** [comment]
Ah, thanks, yes that was it! Apparently I looked into every other parameter besides output_nonlinearity... thanks again!
